### PR TITLE
Handle Generic CPEs for "Running On" Configurations

### PR DESCRIPTION
**Pull request by GabrielEValenzuela**
|Related issue|
|---|
| #23149|

# Objetive
This PR aims to enhance the vulnerability scanner by incorporating logic to accurately handle generic Windows CPEs, especially in "running on" configurations as specified by the NVD. The objective is to improve the scanner's ability to assess vulnerabilities more precisely in environments where specific version details are not available but a generic classification is.

### Changes Implemented
- **Enhanced Logic in `oscpe-global`**:
  - Added logic to handle the template of `oscpe-global.json` to facilitate reverse mapping from platform to vendor using `feed-global.json`.

- **Updates to `osDataCache.hpp`**:
  - Integrated support for new generic CPEs in the cache system to enhance matching capabilities.

- **New Functionality in `scanContext.hpp`**:
  - Implemented `buildCPENameForGenericOrRunning` to construct CPE names based on information received from syscollector messages.

- **Improvements in `packageScanner.hpp`**:
  - Enhanced log messages for better clarity and troubleshooting.
  - Added logic to handle comparisons involving generic CPEs, improving the accuracy of vulnerability assessments.

### Implementation Details
- Modified JSON parsing logic and data structures to support the inclusion of new generic CPEs.
- Updated schema definitions in `oscpe-global.json` to align with the enhanced functionality.
- Conducted extensive unit and integration testing to ensure accurate CPE generation and comparison under various scenarios.

### Testing
- Implemented automated tests to verify the correct functioning of the new logic across different modules (`osDataCache.hpp`, `scanContext.hpp`, and `packageScanner.hpp`).
- Performed regression testing to ensure that existing functionalities are not adversely affected by the new updates.
- Conducted manual testing with simulated syscollector messages to validate the effectiveness of CPE name construction.

### Additional Information
- Documentation has been updated to reflect the new capabilities and guide users on how to leverage the enhanced features effectively.

--------------------------

### CODE REVIEW

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/scanContext.hpp [hunk 1]
```diff
@@ -208,6 +211,58 @@ struct TScanContext final
...
+        // Get the generic CPE from the OS CPE maps
+        std::string cpeTemplate = osCpeMaps.value(KEY_GENERIC_CPE, "");
+        std::string vendorName;
+
+        // Find the vendor name in the reverse vendor maps
+        for (const auto& item : reverseVendorMaps[KEY_REVERSE_VENDOR])
+        {
+            if (item.contains(m_osData.platform))
+            {
+                vendorName = item[m_osData.platform];
```
Comment (sebasfalcone) [inline src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/scanContext.hpp:240]:
```suggestion
                vendorName = item.at(m_osData.platform);
```

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/osDataCache.hpp [hunk 2]
```diff
@@ -26,22 +26,23 @@
...
+    std::string build;               ///< Build of the operating system.
+    std::string platform;            ///< Platform of the operating system.
+    std::string version;             ///< Version of the operating system.
+    std::string release;             ///< Release of the operating system.
+    std::string displayVersion;      ///< Display version of the operating system.
+    std::string sysName;             ///< System name of the operating system.
+    std::string kernelVersion;       ///< Version of the kernel operating system.
+    std::string kernelRelease;       ///< Release of the kernel operating system.
+    std::string cpeName;             ///< CPE name of the operating system.
+    std::string cpeGenericOrRunning; ///< CPE running or generic of the operating system.
```
[the remark this answers is not in the document]
Comment (GabrielEValenzuela) [reply src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/osDataCache.hpp]: Improved it!

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/scanContext.hpp (hunk 1 above)
Comment (GabrielEValenzuela) [reply src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/scanContext.hpp]: Thanks!

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/scanContext.hpp [hunk 3]
```diff
@@ -208,6 +211,57 @@ struct TScanContext final
...
+            m_osData.cpeGeneric.clear();
+            return;
+        }
+
+        // Get the generic CPE from the OS CPE maps
+        std::string cpeTemplate = osCpeMaps.at(KEY_GENERIC_CPE);
+        std::string vendorName;
+
+        // Find the vendor name in the reverse vendor maps
+        for (const auto& item : reverseVendorMaps[KEY_REVERSE_VENDOR])
```
Comment (sebasfalcone) [inline src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/scanContext.hpp:243]: Its preferable to produce an exception than a segfault:
```suggestion
        for (const auto& item : reverseVendorMaps.at(KEY_REVERSE_VENDOR))
```

**Review by sebasfalcone — COMMENTED** [review]
I will take from here @GabrielEValenzuela

**Comment by sebasfalcone** [comment]
## Update
- The target branch changed from 4.8.0 to 4.9.0 
- Still waiting for 4.8.0 to be merged into 4.9.0

**Review by sebasfalcone — CHANGES_REQUESTED** [review]
Good work @GabrielEValenzuela!

Please add an efficacy test for this scenario

**Comment by GabrielEValenzuela** [comment]
# Update

With local changes, test 🟢 

```bash
❯ python -m pytest  -xvv -k "run_process_and_monitor_log14"  wazuh_modules/vulnerability_scanner/qa
==================================================================================== test session starts =====================================================================================
platform linux -- Python 3.11.9, pytest-7.2.2, pluggy-1.5.0 -- /home/gvalenzuela/Documents/Work/wazuh/src/virtual_env/bin/python
cachedir: .pytest_cache
rootdir: /home/gvalenzuela/Documents/Work/wazuh/src
collected 19 items / 18 deselected / 1 selected

wazuh_modules/vulnerability_scanner/qa/test_efficacy_log.py::test_false_negatives[run_process_and_monitor_log14] PASSED                                                                [100%]

============================================================================== 1 passed, 18 deselected in 2.14s ==============================================================================
```

**Review by sebasfalcone — CHANGES_REQUESTED** [review]
Simple, yet effective!

Just one thing, add the documentation for the test 20 ([here is a template](https://github.com/wazuh/wazuh/blob/9721fbeabd6c1c3d2c7de8b14c1af45623302b4e/src/wazuh_modules/vulnerability_scanner/qa/test_data/001/Readme.md))

**Comment by GabrielEValenzuela** [comment]
# Test case II

- After adding the content manually in the code, we run the QA tests again

```bash
❯ python -m pytest --log-cli-level=DEBUG --capture=tee-sys -xvv  wazuh_modules/vulnerability_scanner/qa/
==================================================================================== test session starts =====================================================================================
platform linux -- Python 3.11.9, pytest-7.2.2, pluggy-1.5.0 -- /home/gvalenzuela/Documents/Work/wazuh/src/virtual_env/bin/python
cachedir: .pytest_cache
rootdir: /home/gvalenzuela/Documents/Work/wazuh/src
collected 23 items
[...]
=============================================================================== 23 passed in 240.81s (0:04:00) ===============================================================================
```

**Review by sebasfalcone — APPROVED** [review]
Good work @GabrielEValenzuela!
- Tests 🟢 
- Code 🟢

**Comment by sebasfalcone** [comment]
## Issue blocked
Awaiting for content to be updated in order for QA test to pass in GA

**Review by sebasfalcone — CHANGES_REQUESTED** [review]
Missing efficacy test for the case proposed on the issues:
- https://nvd.nist.gov/vuln/detail/CVE-2012-4142

**Review by sebasfalcone — APPROVED** [review]

**Pull request closed without merging**